iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp

```diff
@@ -1214,17 +1223,72 @@ static LogicalResult createAndPropagateBufferUsedForResultTensor(
   return success();
 }
 
+static LogicalResult createDynamicTemporaryBuffers(
+    Shape::TieShapeOp tieShapeOp, TensorToBufferMap &resultTensorToBufferMap,
+    OpBuilder &builder) {
+  if (resultTensorToBufferMap.count(tieShapeOp.result()) > 0) {
+    // Already has an allocation.
+    return success();
+  }
+
+  auto loc = tieShapeOp.getLoc();
+  RankedTensorType tensorType =
+      tieShapeOp.result().getType().dyn_cast<RankedTensorType>();
+  if (!tensorType) return success();
+  auto shapeType =
+      tieShapeOp.shape().getType().dyn_cast<Shape::RankedShapeType>();
+  if (!shapeType) return success();
+  // TODO(laurenzo): Have a real memory space for temporary buffers?
+  MemRefType memrefType =
+      getMemrefTypeForTensor(tensorType, /*affineMapComposition=*/{},
+                             /*memorySpace=*/0);
+  assert(memrefType.hasRank());
+
+  // Create an allocation at the highest point possible.
+  auto shapeDefOp = tieShapeOp.shape().getDefiningOp();
+  if (shapeDefOp) {
+    builder.setInsertionPointAfter(shapeDefOp);
+  } else {
+    builder.setInsertionPointToStart(builder.getBlock());
+  }
+
+  SmallVector<Value, 4> dynamicDims;
+  for (int i = 0, e = memrefType.getRank(); i < e; ++i) {
+    if (memrefType.isDynamicDim(i)) {
+      auto dim = builder.create<Shape::RankedDimOp>(loc, tieShapeOp.shape(), i);
+      dynamicDims.push_back(dim);
+    }
+  }
+  auto alloc = builder.create<AllocOp>(loc, memrefType, dynamicDims);
```

Comment:
I think that the semantics of std.alloc is more like malloc and so in principle you need to free it somehow (though honestly the upstream behavior isn't terribly well specced). We have std.alloca which is more like a stack object btw, though I think this is one place where yet again we are seeing a phase ordering issue (similar to VMLA lowering) between what fusions can do and what dispatch region partitioning can do. In the VMLA case of course we can paper of it pretty easily since VMLA can allocate random full-size buffers internally to itself, but that's not in general the case for e.g. SPIRV compilation.